additional labels and annotations util

mengqiy opened the pull request:
An implementation of `Transformer` interface to apply additional labels annotations.

mengqiy commented:
Non-test and non-boilerplate code is only about 150 lines :)

monopole commented on pkg/kinflate/util/labelsandannotations_test.go:
```diff
@@ -0,0 +1,320 @@
+	}
+}
+
+var labeledObj1 = unstructured.Unstructured{
```
see commend above, thanks

monopole commented on pkg/kinflate/util/util.go:
```diff
@@ -111,6 +112,53 @@ func Encode(in map[GroupVersionKindName]*unstructured.Unstructured) ([]byte, err
 	return buf.Bytes(), nil
 }
 
+// SelectByGVK returns true if selector is a superset of in; otherwise, false.
```
Please define superset with some examples.

its not obvious to me that if selector is nil, we have a superset condition.  is nil like a wildcard then?

monopole commented on pkg/kinflate/util/util.go:
```diff
@@ -111,6 +112,53 @@ func Encode(in map[GroupVersionKindName]*unstructured.Unstructured) ([]byte, err
+	return true
+}
+
+func CompareMap(m1, m2 map[GroupVersionKindName]*unstructured.Unstructured) error {
```
If CompareMap is a test-only utility it should be in a _test.go file and should be private.

If it is meant to be used outside tests, then it needs its own individual test and godoc.

mengqiy commented:
@monopole Addressed the comments.
PTAL at https://github.com/kubernetes/kubectl/pull/218/files#diff-c1f60fb8ec4a94908d114d943a2fa297R27 to check if the name is good enough :)

monopole commented on pkg/kinflate/util/labelsandannotations.go:
```diff
@@ -22,35 +22,41 @@ import (
-	pathConfigs   []PathConfig
+// MapTransformationOptions contains a map string->string and path configs
+// The map will be applied to the fields specified in path configs.
+type MapTransformationOptions struct {
```
Please call this MapTransformer, or TransformerByMap, or something.  

Its an implementation of Transformer.

monopole commented on pkg/kinflate/util/labelsandannotations.go:
```diff
@@ -22,35 +22,41 @@ import (
-func (o *ApplyAdditionalMapOptions) CompleteForAnnotations(m map[string]string, pathConfigs []PathConfig) {
-	o.additionalMap = m
+// CompleteForAnnotations fills up the MapTransformationOptions for annotations transformation.
+func (o *MapTransformationOptions) CompleteForAnnotations(m map[string]string, pathConfigs []PathConfig) {
```
likewise drop

monopole commented on pkg/kinflate/util/labelsandannotationsconfig.go:
```diff
@@ -20,6 +20,8 @@ import (
 
+// DefaultLabelsPathConfigs is the default configuration for mutating labels and
+// selector fields for native k8s APIs.
 var DefaultLabelsPathConfigs = []PathConfig{
```
It seems to me these can be package private.  They could even be embedded in the constructors suggested above.

monopole commented on pkg/kinflate/util/util.go:
```diff
@@ -112,7 +119,13 @@ func Encode(in map[GroupVersionKindName]*unstructured.Unstructured) ([]byte, err
+// <Group: "extensions", Version: "v1beta1", Kind: "Deployemt">.
+// selector <Group: "apps", Version: "", Kind: "Deployemt"> CANNOT select
+// <Group: "extensions", Version: "v1beta1", Kind: "Deployemt">.
 func SelectByGVK(in schema.GroupVersionKind, selector *schema.GroupVersionKind) bool {
```
thanks much better!

monopole commented on pkg/kinflate/util/labelsandannotations.go:
```diff
@@ -0,0 +1,80 @@
+var _ Transformer = &MapTransformationOptions{}
+
+// CompleteForLabels fills up the MapTransformationOptions for labels transformation.
+func (o *MapTransformationOptions) CompleteForLabels(m map[string]string, pathConfigs []PathConfig) {
```
sorry, previous review somehow dropped a comment here.

This is basically a double setter with a hidden default, and raises alarms for me.

Please drop this, in favor of an object that is immutable behind the Transformer interface:

```
func NewDefaultingLabelsMapTransformer(m map[string]string) (Transformer, err) {
  return NewMapTransformer(DefaultLabelsPathConfigs, m)
}

func NewDefaultingAnnotationsMapTransformer(m map[string]string) (Transformer, err) {
  return NewMapTransformer(DefaultAnnotationsPathConfigs, m)
}

func NewMapTransformer(pc []PathConfig, m map[string]string) (Transformer, err) {
  if (m == nil or is otherwise bad in some way) {
   return nil, errors.New("bad map")
  }
  if (pc == nil or is otherwise bad in some way) {
   return nil, errors.New("bad pathconfigs")
  }
  return &MapTransformer{pc, m}, nil
}
```

mengqiy commented:
Updated. Will squash if you want me to do so.
